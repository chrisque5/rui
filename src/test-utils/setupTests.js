// The below part of the code is to fix the issue
// TypeError: window.matchMedia is not a function
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});
