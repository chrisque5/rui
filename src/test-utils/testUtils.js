/* eslint-disable import/no-extraneous-dependencies, react/prop-types */
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Form } from 'antd';

const middlewares = [thunk];
const renderWithForm = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}><Form>{children}</Form></Provider>;

  return render(ui, { wrapper, ...renderOptions });
};

const renderWithProvider = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}>{children}</Provider>;

  return render(ui, { wrapper, ...renderOptions });
};
// re-export everything
export * from '@testing-library/react';
// override render method
export { renderWithForm, renderWithProvider };
