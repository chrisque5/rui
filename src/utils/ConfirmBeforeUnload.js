import { useEffect, useCallback } from 'react';

const ConfirmBeforeUnload = () => {
  const onBeforeUnload = useCallback((e) => {
    e.preventDefault();
    e.returnValue = '';
  }, []);

  useEffect(() => {
    window.addEventListener('beforeunload', onBeforeUnload);

    return () => {
      window.removeEventListener('beforeunload', onBeforeUnload);
    };
  }, [onBeforeUnload]);

  return null;
};

export default ConfirmBeforeUnload;
