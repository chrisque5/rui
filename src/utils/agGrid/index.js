import {
  approvalsCellRenderer, approvalsSummaryCellRenderer, isUnderInvestigationCellRenderer, stpCellRenderer, stpFilterCellRenderer,
} from './js-renderers';
import {
  dateFormatter, formatDate, numberFormatter, timeFormatter,
  valueDateFormatter, fixingDateFormatter, publishDateFormatter,
  priceFormatter, brokerageFormatter,
} from './formatters';
import { dateFilterComparator, numberFilterComparitor, timeFilterComparator } from './comparitors';
import { getExportMenuItems } from './exportUtils';

export {
  approvalsCellRenderer,
  approvalsSummaryCellRenderer,
  dateFilterComparator,
  dateFormatter,
  formatDate,
  getExportMenuItems,
  isUnderInvestigationCellRenderer,
  numberFilterComparitor,
  numberFormatter,
  stpCellRenderer,
  stpFilterCellRenderer,
  timeFilterComparator,
  timeFormatter,
  valueDateFormatter,
  fixingDateFormatter,
  publishDateFormatter,
  priceFormatter,
  brokerageFormatter,
};
