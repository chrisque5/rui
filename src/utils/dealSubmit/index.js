import mapFormToFwdChain from './mapFormToFwdChain';
import mapFormToNdfChain from './mapFormToNdfChain';
import mapFormToSptChain from './mapFormToSptChain';

export { mapFormToFwdChain, mapFormToNdfChain, mapFormToSptChain };
