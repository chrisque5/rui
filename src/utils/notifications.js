import { notification, Button } from 'antd';
import { CloseOutlined } from '@ant-design/icons';

const clearAll = () => {
  notification.destroy();
};

const closeIcon = (
  <>
    <Button size="small" data-testid="btnCloseAll" className="close-all" onClick={() => clearAll()}>Close All</Button>
    <CloseOutlined />
  </>
);

const showErrorNotification = (message, description, duration) => {
  notification.error({
    message, description, duration: duration || 5, closeIcon,
  });
};

const showSuccessNotification = (message, description, duration) => {
  notification.success({
    message, description, duration: duration || 0, closeIcon,
  });
};

const showWarningNotification = (message, description, duration) => {
  notification.warning({
    message, description, duration: duration || 5, closeIcon,
  });
};

const showInfoNotification = (message, description, duration) => {
  notification.info({
    message, description, duration: duration || 5, closeIcon,
  });
};

export {
  closeIcon,
  showErrorNotification,
  showSuccessNotification,
  showWarningNotification,
  showInfoNotification,
  clearAll,
};
