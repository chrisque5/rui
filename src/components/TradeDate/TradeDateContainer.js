import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { ids } from '../../utils/constants';
import {
  getIsTradeDateSubmitModalVisible, getIsTradeDateEnabled, getUserPreferenceSettingsRatesFeed,
} from '../../utils/selectors';
import { toggleIsTradeDateEnabled } from '../../actions/uiActions';
import TradeDate from './TradeDate';
import TradeDateTime from './TradeDateTime';
import TradeDateEnabled from './TradeDateEnabled';
import TradeDateConfirmModal from './TradeDateConfirmModal';
import { getDateTimeInLocalTZ } from '../../utils/helper';
import { showWarningNotification } from '../../utils/notifications';

const TradeDateContainer = ({
  form, tradeDateOnChange, submitDealWithCustomDate,
}) => {
  const dispatch = useDispatch();
  const [initialTime, setInitialTime] = useState(null);
  const isTradeDateConfirmModalVisible = useSelector(getIsTradeDateSubmitModalVisible);
  const isTradeDateEnabled = useSelector(getIsTradeDateEnabled);
  const isRatesFeedEnabled = useSelector(getUserPreferenceSettingsRatesFeed);

  const onCheckBoxChange = (e) => {
    const checked = e.target.checked;
    dispatch(toggleIsTradeDateEnabled(checked));
    // if checkbox toggled on initialise the time
    if (checked) {
      const currentTime = getDateTimeInLocalTZ();
      setInitialTime(currentTime);
      if (form) {
        form.setFieldsValue({
          [ids.TRADE_DATE]: currentTime,
          [ids.TRADE_TIME]: currentTime,
        });
      }
      if (isRatesFeedEnabled) {
        showWarningNotification(
          'Rate Error',
          'Rate feed is not available when overriding trade date/time. Please set rates manually.',
        );
      }
    } else {
      // if checkbox toggled off reset the time
      setInitialTime(null);
      if (form) {
        form.resetFields([ids.TRADE_DATE, ids.TRADE_TIME]);
      }
      tradeDateOnChange();
    }
  };

  return (
    <>
      <TradeDateEnabled form={form} id={ids.IS_TRADE_DATE_ENABLED} onChange={onCheckBoxChange} />
      <TradeDate
        form={form}
        isTradeDateEnabled={isTradeDateEnabled}
        initialTime={initialTime}
        id={ids.TRADE_DATE}
        tradeDateOnChange={tradeDateOnChange}
      />
      <TradeDateTime
        form={form}
        isTradeTimeEnabled={isTradeDateEnabled}
        initialTime={initialTime}
        id={ids.TRADE_TIME}
      />
      <TradeDateConfirmModal form={form} onAccept={submitDealWithCustomDate} isVisible={isTradeDateConfirmModalVisible} />
    </>
  );
};

TradeDateContainer.propTypes = {
  form: PropTypes.shape().isRequired,
  submitDealWithCustomDate: PropTypes.func.isRequired,
  tradeDateOnChange: PropTypes.func.isRequired,
};

export default TradeDateContainer;
