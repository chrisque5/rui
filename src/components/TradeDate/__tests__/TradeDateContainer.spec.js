import { renderWithForm, waitFor, fireEvent } from 'test-utils/testUtils';
import TradeDateContainer from '../TradeDateContainer';
import { ids } from '../../../utils/constants';
import * as notifications from '../../../utils/notifications';

jest.mock('../../../utils/helper', () => ({ getDateTimeInLocalTZ: jest.fn().mockReturnValue({ momentTz: 1586386800000 }) }));
jest.mock('moment', () => () => jest.requireActual('moment')('2020-06-01T00:00:00.000Z'));

const mockForm = {
  getFieldValue: (field) => {
    if (field === ids.IS_TRADE_DATE_ENABLED) {
      return true;
    }
    return 'dummy';
  },
  resetFields: () => {},
  setFieldsValue: () => jest.fn(),
};

const setFieldsValueSpy = jest.spyOn(mockForm, 'setFieldsValue');
const resetFieldsSpy = jest.spyOn(mockForm, 'resetFields');
const showWarningNotificationSpy = jest.spyOn(notifications, 'showWarningNotification');

const tradeDateOnChange = jest.fn();
const submitDealWithCustomDate = jest.fn();
const defaultProps = { form: mockForm, submitDealWithCustomDate, tradeDateOnChange };
const initialState = {
  user: { preferences: { settings: { ratesFeed: true } } },
  ui: { isTradeDateConfirmModalVisible: false, isTradeDateEnabled: false },
};

describe('<TradeDateContainer />', () => {
  afterEach(() => {
    setFieldsValueSpy.mockClear();
    resetFieldsSpy.mockClear();
    showWarningNotificationSpy.mockClear();
  });

  test('renders snapshot as expected', () => {
    const { container } = renderWithForm(<TradeDateContainer {...defaultProps} />, { initialState });
    expect(container).toMatchSnapshot();
  });

  test('initialises trade date/time when checkbox is enabled', async () => {
    const { getByTestId } = renderWithForm(<TradeDateContainer {...defaultProps} />, { initialState });
    const chkTradeDateEnabled = await waitFor(() => getByTestId(`chk-${ids.IS_TRADE_DATE_ENABLED}`));
    fireEvent.click(chkTradeDateEnabled);
    expect(setFieldsValueSpy).toHaveBeenCalledTimes(1);
    expect(setFieldsValueSpy).toHaveBeenCalledWith({
      [ids.TRADE_DATE]: { momentTz: 1586386800000 },
      [ids.TRADE_TIME]: { momentTz: 1586386800000 },
    });
  });

  test('shows warning message when checkbox is enabled and rates feed enabled', async () => {
    const { getByTestId } = renderWithForm(<TradeDateContainer {...defaultProps} />, { initialState });
    const chkTradeDateEnabled = await waitFor(() => getByTestId(`chk-${ids.IS_TRADE_DATE_ENABLED}`));
    fireEvent.click(chkTradeDateEnabled);
    expect(showWarningNotificationSpy).toHaveBeenCalledTimes(1);
  });

  test('does not show warning message when checkbox is enabled and rates feed disabled', async () => {
    const updatedState = {
      ...initialState,
      user: { preferences: { settings: { ratesFeed: false } } },
    };
    const { getByTestId } = renderWithForm(<TradeDateContainer {...defaultProps} />, { updatedState });
    const chkTradeDateEnabled = await waitFor(() => getByTestId(`chk-${ids.IS_TRADE_DATE_ENABLED}`));
    fireEvent.click(chkTradeDateEnabled);
    expect(showWarningNotificationSpy).toHaveBeenCalledTimes(0);
  });

  test('resets trade date/time when checkbox is disabled', async () => {
    const { getByTestId } = renderWithForm(<TradeDateContainer {...defaultProps} />, { initialState });
    const chkTradeDateEnabled = await waitFor(() => getByTestId(`chk-${ids.IS_TRADE_DATE_ENABLED}`));
    fireEvent.click(chkTradeDateEnabled);
    fireEvent.click(chkTradeDateEnabled);
    expect(resetFieldsSpy).toHaveBeenCalledTimes(1);
    expect(resetFieldsSpy).toHaveBeenCalledWith([ids.TRADE_DATE, ids.TRADE_TIME]);
    expect(tradeDateOnChange).toHaveBeenCalledTimes(1);
  });
});
