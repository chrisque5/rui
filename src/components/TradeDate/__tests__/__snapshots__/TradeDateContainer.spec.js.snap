// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TradeDateContainer /> renders snapshot as expected 1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="ant-row ant-form-item trade-date-enabled"
    >
      <div
        class="ant-col ant-form-item-control"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <label
              class="ant-checkbox-wrapper"
            >
              <span
                class="ant-checkbox"
              >
                <input
                  class="ant-checkbox-input"
                  data-testid="chk-isTradeDateEnabled"
                  id="isTradeDateEnabled"
                  type="checkbox"
                  value=""
                />
                <span
                  class="ant-checkbox-inner"
                />
              </span>
              <span>
                <span
                  class="label-text"
                >
                  Override Trade Date
                </span>
                <span
                  aria-label="info-circle"
                  class="anticon anticon-info-circle"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="info-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm32 664c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V456c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v272zm-32-344a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"
                    />
                  </svg>
                </span>
              </span>
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-row ant-form-item item-trade-date"
    >
      <div
        class="ant-col ant-form-item-control"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <div
              class="ant-picker ant-picker-disabled"
              style="padding-right: 5px; width: 95%;"
            >
              <div
                class="ant-picker-input"
              >
                <input
                  autocomplete="off"
                  disabled=""
                  id="tradeDate"
                  placeholder="Trade Date"
                  readonly=""
                  size="12"
                  title=""
                  value=""
                />
                <span
                  class="ant-picker-suffix"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-row ant-form-item item-trade-time"
    >
      <div
        class="ant-col ant-form-item-control"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <div
              class="ant-picker ant-picker-disabled"
              style="width: 100%;"
            >
              <div
                class="ant-picker-input"
              >
                <input
                  autocomplete="off"
                  disabled=""
                  id="tradeTime"
                  placeholder="Trade Time"
                  readonly=""
                  size="10"
                  title=""
                  value=""
                />
                <span
                  class="ant-picker-suffix"
                >
                  <span
                    aria-label="clock-circle"
                    class="anticon anticon-clock-circle"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="clock-circle"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                      />
                      <path
                        d="M686.7 638.6L544.1 535.5V288c0-4.4-3.6-8-8-8H488c-4.4 0-8 3.6-8 8v275.4c0 2.6 1.2 5 3.3 6.5l165.4 120.6c3.6 2.6 8.6 1.8 11.2-1.7l28.6-39c2.6-3.7 1.8-8.7-1.8-11.2z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
